keps/sig-node/20190226-migrate-in-tree-dockershim-to-out-of-tree.md

```diff
@@ -0,0 +1,148 @@
+
+### Cons
+- Migration pain with a new binary in addition to kubelet.
+- The number of affected users maybe large.
+- Updating all the eco-system tools to support the new cri-dockerd.
+- Many people use the built in dockershim for in-cluster image build. While that may not be something we recommend for a variety of reasons, it will be a breaking change for these users.
+- CRI is still in alpha，should probably get a 1.0 out there splitting out dockershim completely from kubelet.
+- Existed CNI and CSI plugins may also be affected.
+
+> You can check [the discussion in sig-node mailing list](https://groups.google.com/forum/#!msg/kubernetes-sig-node/0qVzfugYhro/l6Au216XAgAJ) for more details. 
+
+Based on all the discussion, we agree that we should not rush to immediate decision. At the same time, it's the right time to start designing and documenting dockershim deprecation criteria and plan, which will be the main content of rest of this KEP.
+
+### Goals
+
+- A concrete dockershim deprecation criteria.
+- A brief plan to deprecate dockershim spanning multiple releases.
+- Issues need to be fixed in kubelet in order to separate dockershim from it.
+
+### Non-Goals
+
+- Deprecation of dockershim immediately without consideration for users and WIP efforts depending on it.
+- Refactoring or re-design of dockershim itself due to deprecation.
+
+## Proposal
+
+### Dockershim deprecation criteria
+
+- CRI itself is beta.
+- kubelet has no dependency on dockershim/docker in its whole lifecycle. 
+- All node related features are CRI generic and have no "back door" dependency on dockershim/docker.
+- Reasonable benchmark result of performance degradation after moving dockershim to out-of-tree.
+- A out-of-tree CRI for docker is implemented and well maintained, and become to beta.
+- E2E test framework has been updated with fully support of out-of-tree CRI container runtime.
+
+### Dockershim deprecation plan
+
+Step 1: Stabilize in-tree dockershim and decouple dockershim from kubelet (but still in-tree).
+
+Target releases: 1.15, 1,16, 1.17
+
+Actions:
+
+- Mark in-tree dockershim as "maintenance mode":
+  - CRI generic changes/features can continue on dockershim.
+  - WIP efforts on dockershim can continue and go to complete.
+  - dockershim/docker specific changes/features should be rejected.
+- Revisit and fix dependency issues of dockershim/docker in kubelet (e.g. #54405). Currently, kubelet still has:
+  - vendored dockershim 
+  - flags that are used to configure dockershim.
+  - support to get container logs when docker uses journald as the driver.
+  - logic of moving docker processes to a given cgroup
+  - TBD anything else?
+- Package in-tree dockershim separately from kubelet and provide a "option" to enable/disable it.
+- Ensure e2e/Node e2e test framework is CRI generic and test cases are independent of container runtime.
+
+Step 2: Work out a out-of-tree CRI for docker
```

Comment:
does this not already exist in the form of containerd?

Reply:
I assume they are different. During recent sig-node meeting, folk from Docker Inc are willing to maintain a CRI-docker.

Reply:
All recent versions of docker are built on top of containerd, so I'm unclear on what functionality dockershim has that cri-containerd does not.

@Random-Liu do you know if there is any reason to use dockershim over cri-containerd with recent versions of docker?

Reply:
I think in the long run, we do expect people to move to CRI conformant container runtimes like containerd and cri-o.

However, people may still need time to make the move, one of the biggest reason is that with dockershim, you can actually see all your Kubernetes containers in Docker, and people's tool may have been relying on that fact. For example, people may have built security tools, monitoring tools, logging tools around that.

With containerd or cri-o, you won't see Kubernetes containers in Docker any more.

Reply:
Another reason is that today's Kubernetes windows support is still based on dockershim. cri-containerd Windows is WIP, but not finished yet.

Reply:
so we cloud remain both mechanisms currently and depreciated the old way gradually.

Reply:
Yes, there are still features that are supported by docker/dockershim, but not with containerd.